File: 0089-gray-code/0089-gray-code.cpp
class Solution {
public:
    
    void solve(int i, int n, vector<int> &v, vector<bool> &vis, vector<int> &ans)
    {
        if(i == (1 << n)){
            if((v.back() & (v.back() - 1)) == 0)
                ans = v;
            return;
        }
        
        if(!ans.empty())
            return;
        
        int num = v.back();
        for(int bit = 0; bit <= 15; bit++)
        {
            int mask = 1 << bit;
            if(num & mask)
            {
                int no = num & ~mask;
                if(no < (1 << n) && !vis[no])
                {
                    vis[no] = 1;
                    v.push_back(no);
                    solve(i + 1, n, v, vis, ans);
                    v.pop_back();
                    vis[no] = 0;
                }
            }
            else if((num & mask) == 0)
            {
                int no = num | mask;
                if(no < (1 << n) && !vis[no])
                {
                    vis[no] = 1;
                    v.push_back(no);
                    solve(i + 1, n, v, vis, ans);
                    v.pop_back();
                    vis[no] = 0;
                }
            }
        }
    }
    
    vector<int> grayCode(int n) {
        vector<int> v = {0}, ans;
        vector<bool> vis(1 << n);
        vis[0] = 1;
        solve(1, n, v, vis, ans);
        return ans;
    }
};

File: 0089-gray-code/0089-gray-code.go
func solve(i int, n *int, vis *[]int, tmp *[]int, ans *[]int) {
    if i == (1 << *n) {
        val := (*tmp)[len(*tmp) - 1]
        if val & (val - 1) == 0 {
            *ans = append((*ans)[:0], (*tmp)...)
        }
        return
    }
    
    if len(*ans) > 0 {
        return
    }
    
    for j := 0; j <= 15; j++ {
        num, mask := (*tmp)[len(*tmp) - 1], 1 << j
        if num & mask > 0 {
            num &= ^mask
        } else {
            num |= mask
        }
        if num < (1 << *n) && (*vis)[num] == 0 {
            (*vis)[num] = 1
            *tmp = append(*tmp, num)
            solve(i + 1, n, vis, tmp, ans)
            *tmp = (*tmp)[:len(*tmp) - 1]
            (*vis)[num] = 0
        }
    }
}

func grayCode(n int) []int {
    vis, tmp, ans := make([]int, 1 << n), []int{0}, []int{}
    vis[0] = 1
    solve(1, &n, &vis, &tmp, &ans)
    return ans
}